flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamSource.java

```diff
@@ -214,14 +216,26 @@ public AutomaticWatermarkContext(
 			this.watermarkInterval = watermarkInterval;
 			this.reuse = new StreamRecord<T>(null);
 
+			// if it is a source, then we cast and cache it
+			// here so that we do not have to do it in every collect(),
+			// collectWithTimestamp() and emitWatermark()
+
+			if (!(owner instanceof AsyncExceptionChecker)) {
+				throw new IllegalStateException("The ManualWatermarkContext can only be used " +
+					"with sources that implement the AsyncExceptionThrower interface.");
+			}
+			this.source = (AsyncExceptionChecker) owner;
+
 			long now = owner.getCurrentProcessingTime();
 			this.watermarkTimer = owner.registerTimer(now + watermarkInterval,
 				new WatermarkEmittingTask(owner, lockingObjectParam, outputParam));
 		}
 
 		@Override
 		public void collect(T element) {
-			owner.checkAsyncException();
+			if (source != null) {
```

Comment:
I don't think these null checks are needed, because the `IllegalStateException` is thrown if `owner` is `null`.